### PR TITLE
Try improve performance of aggregation in order of sorting key

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve performance of aggregation in order of sorting key (with enabled setting `optimize_aggregation_in_order`)

### CODE REVIEW

**Comment by CurtizJ** [comment]
Fuzzer report is related.

File: src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * The second step of aggregation in order of sorting key.
+ * The transform receives k inputs with partially aggregated data,
+ * sorted by group by key (prefix of sorting key).
+ * Then it merges aggregated data from inputs by the following algorithm:
+ *  - At each step find the smallest value X of the sorting key among last rows of current blocks of inputs.
+ *    Since the data is sorted in order of sorting key and has no duplicates (because of aggregation),
+ *    X will never appear later in any of input streams.
+ *  - Aggregate all rows in current blocks of inputs up to the upper_bound of X using
+ *    regular hash table algorithm (Aggregator::mergeBlock).
```
Comment (akuzm) [inline src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h:23]: I don't understand this part -- if the input data is sorted by the `GROUP BY` key, and we cut the parts of input blocks with same keys, and pass them to `mergeBlock` -- the hash table is always going to have only one element? Would it make sense to just use one aggregation state?

**Comment by akuzm** [comment]
The performance improvement in `read_in_order_many_parts` is nice. For `aggregation_in_order`, the improvement is only noticeable for the query with `GROUP BY CounterID, EventDate, intHash32(UserID)`, and not for other queries. Is this expected?

https://clickhouse-test-reports.s3.yandex.net/19401/2e8b45afc17d4bce41f6d5517a49842a15d69720/performance_comparison/all-queries.html#all-query-times.aggregation_in_order.0

**Comment by CurtizJ** [comment]
But in the test `read_in_order_many_parts` query ```SELECT sum(val2) FROM mt_100_parts GROUP BY val1 FORMAT Null``` has speeded up. For `aggregation_in_order` it's expected, that there is almost no improvement, because `hits` table contains only one or few parts and merge stage takes non-significant amount of time.

File: src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h (hunk 1 above)
Comment (CurtizJ) [reply src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h:23]: There can be duplicates among all streams for one range. The hash table at one step will contain all keys <= X from all blocks, but not one element. There are already algorithm (`AggregatingSortedAlgorithm`), that merges and aggregates sorted data and uses one aggregation state. It is a simple k-way merge algorithm and it makes `O(n*log(k))` comparisons, where n -- number of rows, k -- number of parts. Algorithm from this PR makes `O(n + k * log(n)) ` operations, n -- for hash table, `k * log(n)` -- for finding positions in blocks. It is better than `O(n*log(k))`, when k is not too big and not too small (about 100-1000).

Comment (akuzm) [reply src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h:23]: Thanks, I think I understand now.

Comment (akuzm) [reply src/Processors/Merges/Algorithms/FinishAggregatingInOrderAlgorithm.h:23]: Maybe we should just copy-paste this explanation into the comment as well? It mentions the other algorithm and discusses the complexity, I think this is a useful comment.

**Comment by akuzm** [comment]
Integration and perf test failures are unrelated.

**Pull request merged**